Automattic/jetpack
# CRM: Fix quote placeholders on the quote template, client portal, pdf and emails

## Proposed changes:

* This PR focuses on fixing Quote ID, URL and Edit URL quote placeholders when used in quote templates as well as email templates (quote accepted and proposal notification), and also where displayed in the Client Portal and via PDF.
* It is also the base PR upon which two other PRs with further placeholder fixes are based (see below). Because of that, there are some general 'additional notes' below related to other uncovered placeholders.
* The Quote ID placeholder as well as Quote URL are now not rendered in the Quote Editor when added to a Quote Template, but they are rendered elsewhere (Client Portal, PDF, Quote accepted and proposal notification mail templates).
* The Quote Edit URL placeholder won't display anything in the Quote Editor if added to a Quote Template, nor will it display in the Client Portal or PDF (as a link to edit the Quote URL doesn't make sense to share to a client). Similarly it won't display in the quote notification email, however it will show correctly in the quote accepted email notification. 

There are two further PRs based on this PR:
- https://github.com/Automattic/jetpack/pull/34569
- https://github.com/Automattic/jetpack/pull/34568

Related P2 post: pbhBOz-4fT-p2

## Additional notes:

- I've chosen to leave Quote checkbox and autonumber custom fields as 'won'tfix' (they work in mail templates, not in the quote editor, client portal or pdf). The reason they don't work in the quote editor and therefore preview and pdf is:
  - for the checkbox the relevant checkbox names are different for each checkbox within a group, and that not matching the parent group name (number additions to the names). Solving that would like involve  creating an additional hidden input so the name matches the custom field name, and setting it's value to also be the same as all checked checkbox values via JS. 
  - For autonumber custom fields, these are added after the quote is initially generated so there is a similar problem as with quote ID and quote URL - however as it's a custom field we can't know the custom field name in advance adding extra complication. 
  - In both cases, how likely are people to use these? Hence it seems a wontfix given the complication imo (but as with all placeholder availability we should clarify where it works), and I can also add a closed GH issue just for searchability.
- For the same reason, I've left adding currency formatting and adding a currency type to the quote price custom field (also in mail templates - custom price fields aren't showing a currency value)

## Next Steps

Once all relevant PRs are merged, then I'll document which placeholders work where on a comment on this issue: https://github.com/Automattic/zero-bs-crm/issues/3005
The Quote Template placeholder support doc also needs updating: https://github.com/Automattic/jedi-jetpack-enhancements-of-docs-for-ai/issues/327


### Other information:


## Jetpack product discussion

There are several related issues:
- https://github.com/Automattic/zero-bs-crm/issues/3004 (a parent issue of sorts, so this issue will not close that)
- https://github.com/Automattic/zero-bs-crm/issues/1601
- https://github.com/Automattic/zero-bs-crm/issues/3060

As a bonus, this PR will also fix the following issues:
- https://github.com/Automattic/zero-bs-crm/issues/3384

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:


* To test, check out this PR locally or via Jurassic Ninja and the Jetpack Beta plugin.
* In a new Quote Template, add the following placeholders: `##QUOTE-ID##`, `##QUOTE-EDIT-URL##` and `##QUOTE-URL##`. Save that template.
* Create a new quote and both before and after saving notice that both the Quote ID and Quote URL placeholders stay unrendered, whereas the Quote Edit URL placeholder is removed.
* After clicking save, preview the quote in the Client Portal. All fields except the for the Quote Edit URL should display correctly.
* Then download the PDF for the quote, and the same should be visible - all but the Quote Edit URL.
* To test mail templates - `/wp-admin/admin.php?page=zbs-email-template` - test the quote accepted email template by accepting the quote you've just created in the client portal and viewing the admin email. Alternatively, add an error log before the final `$html` return in `zeroBSCRM_quote_generateAcceptNotifHTML()` and check the result. You should notice both the Quote Edit URL and Quote URL should be visible. You can add the Quote ID placeholder and that should also display.
* To test the quote proposal email, add the Quote URL and Quote ID placeholders to the 'You have received a new proposal' template - `/wp-admin/admin.php?page=zbs-email-templates&zbs_template_id=4`. Test this by sending the new quote via email in teh quote editor, or by adding an error_log before the `$html` is returned in `zeroBSCRM_quote_generateNotificationHTML()`.
